File: go/hamming/hamming.go
// Package hamming provides the distance of two protein strands
package hamming

import (
	e "errors"
	s "strings"
)

// Distance takes two protein strands and determines their "distance" apart.
func Distance(a, b string) (int, error) {
	if len(a) != len(b) {
		return 0, e.New("invalid strings")
	}

	aChars := s.Split(a, "")
	bChars := s.Split(b, "")
	distance := 0

	for index, letter := range aChars {
		if letter != bChars[index] {
			distance = distance + 1
		}
	}

	return distance, nil
}

File: scala/hamming/src/main/scala/Hamming.scala
object Hamming {
  def distance(a: String, b: String): Option[Int] = {
    if (a.length != b.length) {
      None
    } else {
      Some(a.zip(b).filter { case (x, y) => x != y }.length)
    }
  }
}
